File: backend/src/schemas/feedback.py
"""
Pydantic schemas for feedback API request/response validation.
"""

from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class FeedbackCreate(BaseModel):
    """Schema for creating a new feedback entry."""

    type: Literal["bug", "suggestion", "general"] = Field(
        ..., description="Feedback type"
    )
    title: str = Field(..., min_length=1, max_length=200, description="Short title")
    body: str = Field(..., min_length=10, max_length=5000, description="Feedback content")
    page_url: Optional[str] = Field(None, max_length=500, description="Page URL")
    context: Optional[Dict[str, Any]] = Field(None, description="Optional context")

    model_config = {"extra": "forbid"}


class FeedbackResponse(BaseModel):
    """Schema for feedback in API responses."""

    id: str
    user_id: str
    type: str
    title: str
    body: str
    page_url: Optional[str] = None
    context: Optional[Dict[str, Any]] = None
    status: str
    admin_notes: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    submitter_email: Optional[str] = None  # Populated when listing with user loaded

    model_config = {"from_attributes": True}


class FeedbackUpdate(BaseModel):
    """Schema for admin updates to feedback (status, admin_notes)."""

    status: Optional[Literal["open", "in_progress", "resolved", "closed"]] = None
    admin_notes: Optional[str] = None

    model_config = {"extra": "forbid"}


class FeedbackListResponse(BaseModel):
    """Schema for paginated list of feedback."""

    feedbacks: List[FeedbackResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/types/feedback.ts
/**
 * Types for the feedback feature: submission, list response, and admin stats.
 */

export type FeedbackType = "bug" | "suggestion" | "general";
export type FeedbackStatus =
  | "open"
  | "in_progress"
  | "resolved"
  | "closed";

export interface FeedbackCreate {
  type: FeedbackType;
  title: string;
  body: string;
  page_url?: string;
  context?: Record<string, unknown>;
}

export interface Feedback {
  id: string;
  user_id: string;
  type: FeedbackType;
  title: string;
  body: string;
  page_url?: string;
  context?: Record<string, unknown>;
  status: FeedbackStatus;
  admin_notes?: string;
  created_at: string;
  updated_at?: string;
  submitter_email?: string;
}

export interface FeedbackListResponse {
  feedbacks: Feedback[];
  total: number;
  page: number;
  page_size: number;
}

export interface FeedbackStats {
  by_type: Record<string, number>;
  by_status: Record<string, number>;
  total: number;
}
